File: Rust/Benchmark/src/fibonacci.rs
// fibonacci.rs
// Calulate Fibonacci numbers in Rust
//
// vim: ft=rust sw=4 ts=4
//

use num::bigint::BigInt;
use num::{One, Zero};

// a naive recursive implementation
pub fn fib_naive(n: u64) -> u64 {
    if n < 2 {
        n
    } else {
        fib_naive(n - 1) + fib_naive(n - 2)
    }
}

// optimzed tail-recursive implmentation
pub fn fib(n: u64) -> BigInt {
    fib_loop(n, BigInt::zero(), BigInt::one())
}
fn fib_loop(n: u64, a: BigInt, b: BigInt) -> BigInt {
    if n == 0 {
        return a;
    }
    fib_loop(n - 1, b.clone(), a + b)
}

// unit tests
#[cfg(test)]
mod test {
    use crate::fibonacci::{fib, fib_naive};

    #[test]
    fn test_fib_naive() {
        assert_eq!(fib_naive(0), 0);
        assert_eq!(fib_naive(1), 1);
        assert_eq!(fib_naive(2), 1);
        assert_eq!(fib_naive(3), 2);
        assert_eq!(fib_naive(4), 3);
        assert_eq!(fib_naive(10), 55);
        assert_eq!(fib_naive(20), 6765);
    }

    #[test]
    fn test_fib() {
        use num::bigint::ToBigInt;
        assert_eq!(fib(0), 0.to_bigint().unwrap());
        assert_eq!(fib(1), 1.to_bigint().unwrap());
        assert_eq!(fib(2), 1.to_bigint().unwrap());
        assert_eq!(fib(3), 2.to_bigint().unwrap());
        assert_eq!(fib(4), 3.to_bigint().unwrap());
        assert_eq!(fib(10), 55.to_bigint().unwrap());
        assert_eq!(fib(20), 6765.to_bigint().unwrap());
    }
}

File: OCaml/Benchmark/lib/fibonacci.ml
(* vim: set ft=ocaml sw=2 ts=2: *)

(* fibonacci.ml
 * A package to calculate Fibonacci numbers
 *)

(* naive recursive implementation *)
let rec fib_naive n =
  if n < 2 then n else fib_naive (n-1) + fib_naive (n-2)

(*$= fib_naive & ~printer:string_of_int
  (fib_naive 0) 0
  (fib_naive 1) 1
  (fib_naive 2) 1
  (fib_naive 3) 2
  (fib_naive 4) 3
  (fib_naive 10) 55
  (fib_naive 20) 6765
  (fib_naive 30) 832040
*)


(* optimized recursive implementation *)
let fib n =
  let rec fib_aux n a b =
    if n = 0 then a else fib_aux (n-1) b (Big_int.add_big_int a b)
  in
  fib_aux n Big_int.zero_big_int Big_int.unit_big_int

(*$T fib
  (Big_int.eq_big_int (fib 0) (Big_int.big_int_of_int 0))
  (Big_int.eq_big_int (fib 1) (Big_int.big_int_of_int 1))
  (Big_int.eq_big_int (fib 2) (Big_int.big_int_of_int 1))
  (Big_int.eq_big_int (fib 3) (Big_int.big_int_of_int 2))
  (Big_int.eq_big_int (fib 4) (Big_int.big_int_of_int 3))
  (Big_int.eq_big_int (fib 10) (Big_int.big_int_of_int 55))
  (Big_int.eq_big_int (fib 20) (Big_int.big_int_of_int 6765))
  (Big_int.eq_big_int (fib 30) (Big_int.big_int_of_int 832040))
*)
